Use DataViews for all domains list

merkushin opened the pull request:
Related to #96674 

This solution doesn't address one main point we wanted to have in the beginning and also received as feedback: we don't have all the needed data immediately from the backend.

It is a trade-off for the first iteration.

Implementation of the needed backend has its challenges and will be addressed later: pdxWSz-23k-p2

## Proposed Changes

* Introduce DataViews for all domains list.
* Add `calypso/domains-dataviews` feature flag.

## Why are these changes being made?

* Use the standard approach to render a list on WordPress.

## Testing Instructions

* Turn on feature flags in your browser's console: `window.sessionStorage.setItem('flags',  'calypso/domains-dataviews'])`
* Go to http://calypso.localhost:3000/domains/manage
* Check filtering, sorting, actions, bulk actions.
    * For bulk actions, you need to have at least two domains you own. 
* Click on a domain and check if it opened a pane on the right.
* Switch the tab on the right to Email.
* Click on another domain and make sure the same tab (Email) opened for it.
* Now disable DataViews: `window.sessionStorage.removeItem('flags');` and test everything works with the previous implementation.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

merkushin commented on packages/domains-table/src/utils/get-domain-type-text.tsx:
```diff
@@ -1,9 +1,15 @@
+
 export function getDomainTypeText(
-	domain: ResponseDomain,
+	domain: PartialDomainInfo,
```
This change is not needed anymore, as I changed new code.
However, it might be worth to use `Pick` here.

merkushin commented on client/my-sites/domains/domain-management/dataviews/use-domains.ts:
```diff
@@ -0,0 +1,71 @@
+
+export type QueryDomainExtendedData = QueryDomainData & DomainDataExtension;
+
+export type DomainData = {
```
I'm going to revise this model one more time, using ResponseDomain as the main object and add a few fields to it (from DomainDataExtension).

bogiii commented on client/my-sites/domains/domain-management/dataviews/use-context.ts:
```diff
@@ -0,0 +1,227 @@
+	isAllSitesView: boolean;
+	domainStatusPurchaseActions?: DomainStatusPurchaseActions;
+	onDomainAction?: OnDomainAction;
+	userCanSetPrimaryDomains?: boolean;
+	hideCheckbox?: boolean;
+	isLoadingDomains?: boolean;
+	useMobileCards?: boolean;
+	sidebarMode?: boolean;
+	selectedDomainName?: string;
+	createBulkAction?: ( variables: BulkUpdateVariables ) => Promise< void >;
+	fetchBulkActionStatus?: () => Promise< BulkDomainUpdateStatusQueryFnData >;
+	deleteBulkActionStatus?: () => Promise< void >;
+	currentUserCanBulkUpdateContactInfo?: boolean;
+}
+
+export type DomainsDataViewsProps =
+	| ( BaseDomainsDataViewsProps & { isAllSitesView: true } )
+	| ( BaseDomainsDataViewsProps & { isAllSitesView: false; siteSlug: string | null } );
+
+type Value = {
+	sites: Record< number, SiteExcerptData >;
+	isLoadingSites?: boolean;
+	isLoadingDomains?: boolean;
+	createBulkAction?: ( variables: BulkUpdateVariables ) => Promise< void >;
+	fetchBulkActionStatus?: () => Promise< BulkDomainUpdateStatusQueryFnData >;
+	deleteBulkActionStatus?: () => Promise< void >;
+	isAllSitesView: boolean;
+	domainStatusPurchaseActions?: DomainStatusPurchaseActions;
+	domainsRequiringAttention?: number;
+	handleAutoRenew: ( domains: DomainData[], enable: boolean ) => void;
+	handleUpdateContactInfo: ( domains: DomainData[] ) => void;
+	onDomainsRequiringAttentionChange: ( domainsRequiringAttention: number ) => void;
+	completedJobs: JobStatus[];
+	domainResults: Map< string, DomainUpdateStatus[] >;
+	handleRestartDomainStatusPolling: () => void;
+	onDomainAction( ...parameters: Parameters< OnDomainAction > ): void;
+	updatingDomain: DomainsDataViewsUpdatingDomain | null;
+	userCanSetPrimaryDomains: BaseDomainsDataViewsProps[ 'userCanSetPrimaryDomains' ];
+	isCompact: boolean;
+};
```
I'm suggesting moving the type definition to a separate file.

merkushin replied:
Extracted all types into separated file: client/my-sites/domains/domain-management/dataviews/types.ts

But, honestly, I'm not sure if it was a good idea. Some types just weren't supposed to be exposed...

oandregal commented on client/my-sites/domains/domain-management/dataviews/style.scss:
```diff
@@ -0,0 +1,234 @@
+					margin: auto;
+
+					.dataviews-view-list {
+						.dataviews-view-list__media-wrapper {
```
We should remove all this CSS targeting DataViews internal component.

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/3b267fdac1be6f7a1f9e87eb84260cf3eaeaa7ce

oandregal commented on client/my-sites/domains/domain-management/dataviews/domains-dataviews.tsx:
```diff
@@ -0,0 +1,62 @@
+
+	const layout = sidebarMode ? { list: {} } : { table: {} };
+	return (
+		<div className="domains-dataviews">
```
Can dataviews be updated to the latest version in dotcom? All of this would be simplified by using the latest (see `onClickItem`/`isItemClickable` in [the docs](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#properties)).

merkushin replied:
Yes, updating the library could make things even more easier. 
But I'm afraid it's not safe since the library is used in other parts of the app.
Shouldn't it be a separate task? Maybe even outside the domain-related project?

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-fields.tsx:
```diff
@@ -0,0 +1,127 @@
+				enableHiding: false,
+				enableSorting: true,
+				getValue: ( { item }: { item: DomainData } ) => item.processed.sslStatus,
+				render: ( { item }: { item: DomainData } ) => <SslStatusField domain={ item } />,
```
I haven't checked all render functions here, but this `SslStatusField` component renders a `td`. This is problematic. When DataViews is using the table layout it's going to render a `td` within a `td`; when DataViews is using the list layout it's going to render a `td` within a `li` component.

My recommendation would be not using anything from `domains-table`, which seems to be tailored for something that DataViews already does (creating a table). Instead, we should have specific render functions for each of these fields.

merkushin replied:
![CleanShot 2024-12-26 at 21 05 21@2x](https://github.com/user-attachments/assets/afc7151a-1bc1-4025-a49b-6ba5a8f9db52)

Hum, not sure about `td` here. Earlier, I made [changes](https://github.com/Automattic/wp-calypso/pull/97677/files#diff-6a092857c761a530cb8c8faba5d69bafd1d3b8dfc4863df4430b9f8dbd57f21dR11) for the cell to be wrapped in a `div` when it is needed. (A similar approach was already used for another cell.)

oandregal commented:
I've left some comments for the major things I see and also shared some testing with real data privately. I'm happy to do a second round of review when these are addressed.

ouikhuan commented on client/my-sites/domains/domain-management/dataviews/use-fields.tsx:
```diff
@@ -0,0 +1,77 @@
+import { DomainData } from './use-domains';
+
+export function useFields() {
+	const { __ } = useI18n();
```
Any reason to not using `useTranslate` hook here instead?

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/fb95f90d48d6415c7d4591a78cc1520990bb7bc2

merkushin committed:
Add domains DataViews feature flag

merkushin committed:
Make argument type more specific

merkushin committed:
Allow render the component as a div when needed

merkushin committed:
Add domains DataViews implementation

merkushin committed:
Use domains DataViews in BulkAllDomains with a feature flag

Imran92 commented on client/my-sites/domains/domain-management/dataviews/domains-dataviews.tsx:
```diff
@@ -0,0 +1,77 @@
+
+	useEffect( () => {
+		const fields = getFieldsByBreakpoint( isDesktop, sidebarMode );
+		const fieldsForBreakpoint = [ ...fields ].sort().toString();
```
Looks like the `fields` variable should be named `fieldsForBreakpoint` instead, and `fieldsForBreakpoint` can be called `sortedBreakpointFields` or something like that.

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/d283acb1169e89cdfb44c7a091977bc3dd5d3853

Imran92 commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,313 @@
+						/>
+					</div>
+				);
+			},
```
Should we extract it to a separate function?

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/40b2387ef2706418303cab984fef790fe37fdae9

Imran92 commented on client/my-sites/domains/domain-management/dataviews/types.ts:
```diff
@@ -0,0 +1,103 @@
+	domainsRequiringAttention?: number;
+	handleAutoRenew: ( domains: DomainData[], enable: boolean ) => void;
+	handleUpdateContactInfo: ( domains: DomainData[] ) => void;
+	onDomainsRequiringAttentionChange: ( domainsRequiringAttention: number ) => void;
+	completedJobs: JobStatus[];
+	domainResults: Map< string, DomainUpdateStatus[] >;
+	handleRestartDomainStatusPolling: () => void;
+	onDomainAction( ...parameters: Parameters< OnDomainAction > ): void;
+	updatingDomain: DomainsDataViewsUpdatingDomain | null;
+	userCanSetPrimaryDomains: BaseDomainsDataViewsProps[ 'userCanSetPrimaryDomains' ];
+	isDesktop: boolean;
+	selectedFeature?: string;
+}
+
+type DomainDataExtension = {
+	site_slug: string;
+	domain_status: {
+		status: string;
+		status_type: string;
+		status_weight: number;
+	};
+	can_manage_dns_records: boolean;
+};
+
+export type QueryDomainExtendedData = QueryDomainData & DomainDataExtension;
+
+export type DomainData = {
+	original: QueryDomainExtendedData;
+	processed: ResponseDomain;
+};
+
+export interface DomainsQueryArgs {
+	no_wpcom?: boolean;
+	resolve_status?: boolean;
+	extended_data?: boolean;
+}
+
+export interface DomainsQueryFnData {
+	domains: QueryDomainExtendedData[];
+}
```
Just an optional suggestion, maybe adding some jsdoc comments to explain parts of these will be helpful.

merkushin replied:
Added comments here: https://github.com/Automattic/wp-calypso/pull/97677/commits/5ec52bc6710c2c9461f89e1021a38df230264ca8
If you have an idea how to improve them, please let me know.
Initially, tried to add more comprehensive JSDoc, but then TypeScript liner started complaining I should use types :)

Imran92 reviewed (commented):
Overall it looks good to me. I've found just a couple of issues

If you are not on the first page, and select a domain, the flyout opens, but in the domains list on left side, it jumps to the first page.

Also, as a result of this, the domain does not show or remain as selected on the left side-

https://github.com/user-attachments/assets/32991dda-11c6-4ec4-bcfc-bc32f99c223c

Imran92 commented on client/my-sites/domains/domain-management/dataviews/domains-dataviews.tsx:
```diff
@@ -0,0 +1,77 @@
+	selectedDomainName,
+	openDomainPane,
+}: Props ) => {
+	const { __ } = useI18n();
```
Should we use `useTranslate` here too?

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/bf8860060b154ed44b272745308ba7343ccdf698

Imran92 reviewed (commented):
I've noticed another issue where the selection is acting a bit strange 🤔 

https://github.com/user-attachments/assets/698c6213-c518-49ea-9dfb-4bbabeccdfdb

merkushin committed:
Extract auto-renew dialog into a separate component

merkushin committed:
Refactor variable names

merkushin committed:
Add short descriptions for types

merkushin committed:
Use useTranslate instead of useI18n

merkushin commented:
@Imran92 
> I've noticed another issue where the selection is acting a bit strange 🤔

Hum, not sure what we can do here. This is behavior from DataViews.
The only thing that comes to my mind is to update the library and check it there.
But please tell me if I didn't interpret your message (and video) properly.

Imran92 commented:
Thanks for fixing the ones you've mentioned. Have you worked on this one though? -

> If you are not on the first page, and select a domain, the flyout opens, but in the domains list on left side, it jumps to the first page.
> 
> Also, as a result of this, the domain does not show or remain as selected on the left side-

jorgefilipecosta reviewed (commented):
I noticed the "Renew now" actions which previously was available on a .blog test domain, now is not available.

jorgefilipecosta reviewed (commented):
The owner is empty on all the domains and previously it had values, the same for SSL status.

jorgefilipecosta commented:
Status is showing an error for some domains that previously had a green active status e.g.: a8c.tv.

jorgefilipecosta reviewed (commented):
If I select two domains and try to bulk edit contact information I go to a page "Editing contact info for 1 domain:" and just lists one domain instead of the two domains I selected.

jorgefilipecosta reviewed (commented):
The manage DNS action is not appearing on cases where it previously appeared.

jorgefilipecosta commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/domain-status-field.tsx:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	const domainStatus = domain
+		? resolveDomainStatus( domain.processed, {
```
Could we use item.original.domain_status.status as the domain status? It seems that we are calculating the status in the component responsible for rendering it, which isn't ideal. Ideally, this component should focus on displaying the status with appropriate styling and icons, without performing any computations.

merkushin replied:
Unfortunately, we can't at the moment. Maybe when we switch to a custom backend.

jorgefilipecosta commented:
When I apply the "Manage Auto Renew" bulk action, the toolbar becomes invisible due to the notice. 

<img width="1952" alt="Screenshot 2025-01-02 at 18 01 18" src="https://github.com/user-attachments/assets/cf0983f7-05dc-4755-ac18-99e67d77ce24" />


With the toolbar hidden, it's impossible to unselect items until the notice is dismissed

jorgefilipecosta commented on client/my-sites/domains/domain-management/dataviews/components/bulk-update-notice.tsx:
```diff
@@ -0,0 +1,68 @@
+					status="is-success"
+					onDismissClick={ () => handleDismissNotice( job.id ) }
+				>
+					{ translate( 'Bulk domain updates finished successfully ' ) }
```
Nitpick: 'Bulk domain updates finished successfully ' -> 'Bulk domain updates finished successfully.'

merkushin replied:
Addressed here: https://github.com/Automattic/wp-calypso/pull/98457

jorgefilipecosta commented on client/my-sites/domains/domain-management/dataviews/components/bulk-update-notice.tsx:
```diff
@@ -0,0 +1,68 @@
+					status="is-success"
+					onDismissClick={ () => handleDismissNotice( job.id ) }
+				>
+					{ translate( 'Bulk domain updates finished successfully ' ) }
```
This message also appears when the auto-renew setting for a single domain is changed. I believe it should be conditional and display a different message when only one domain is modified.

merkushin replied:
Addressed here: https://github.com/Automattic/wp-calypso/pull/98457

Imran92 reviewed (commented):
When we paginate in `wordpress.com/sites`, the URL changes to reflect the current page. So if we reload, the same page opens. But it doesn't happen in the `wordpress.com/domains/manage` page

merkushin committed:
Use original backend to load data

merkushin committed:
Fix imported function name

merkushin commented:
> When we paginate in `wordpress.com/sites`, the URL changes to reflect the current page. So if we reload, the same page opens. But it doesn't happen in the `wordpress.com/domains/manage` page

Addressed in a separate PR here: https://github.com/Automattic/wp-calypso/pull/97953

merkushin commented:
> When I apply the "Manage Auto Renew" bulk action, the toolbar becomes invisible due to the notice.
> 
> <img alt="Screenshot 2025-01-02 at 18 01 18" width="1952" src="https://private-user-images.githubusercontent.com/11271197/399773534-cf0983f7-05dc-4755-ac18-99e67d77ce24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcwMDI0NTAsIm5iZiI6MTczNzAwMjE1MCwicGF0aCI6Ii8xMTI3MTE5Ny8zOTk3NzM1MzQtY2YwOTgzZjctMDVkYy00NzU1LWFjMTgtOTllNjdkNzdjZTI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE2VDA0MzU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyYjNlMzc0NjU0NTg2YjkwNmViZmNhYjE2NjU1MTZiY2Y4MTJiOTU5YWY2Y2U5ZDllZTM2NjJhNzFhZGJjMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.l_FEgRgIQVoP52vrN917NYoJLzJdDlC81I3vpMva9_4">
> With the toolbar hidden, it's impossible to unselect items until the notice is dismissed

Addressed in a separate PR here: https://github.com/Automattic/wp-calypso/pull/98457

merkushin commented:
Hum, for some reason GitHub shows failed checks for the merged branch.
When I switch the branch to the proper one, I see green checks.

merkushin committed:
Modify doc comment

merkushin committed:
Handle pagination for Domains DataViews (#97953)

* Store pagination in query params and apply it
* Store sorting params in query parameters
* Set default per page to 50, same as for sites
* Display loader for cells with lazy loading (#98456)
* Bulk actions notification improvements (#98457)

merkushin committed:
Pass hasConnectableSites and use it for DomainsTableSiteCell

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-fields.tsx:
```diff
@@ -0,0 +1,152 @@
+					}
+					// Removes the username that appears in parentheses after the owner's name.
+					// Uses $ and the negative lookahead assertion (?!.*\() to ensure we only match the very last parenthetical.
+					return domain.owner.replace( / \((?!.*\().+\)$/, '' );
```
The render function checks that `domain.owner` exists but this doesn't. Which one is correct? Is this missing a check or the check is redundant in the render? Trying to understand if `domain.owner.replace` can throw an error here.

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/91bececa44af9c57eda117e22f2c27510f60e08d

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-fields.tsx:
```diff
@@ -0,0 +1,152 @@
+			},
+		],
+		[
+			domainsRequiringAttention,
```
`hasConnectableSites` is also a dependency we have to list here.

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/91bececa44af9c57eda117e22f2c27510f60e08d

oandregal commented on client/my-sites/domains/domain-management/dataviews/domains-dataviews.tsx:
```diff
@@ -0,0 +1,103 @@
+	const { isDesktop } = useDomainsDataViewsContext();
+	const queryParams = useQueryParams();
+
+	const { view, setView } = useView( { sidebarMode, isDesktop, queryParams } );
```
This hook is just a wrapper for the useState's initializer function. It'd be clearer if we do:

```js
const [ view, setView ] = useState( () => initializeState( sidebarMode, isDesktop, queryParams ) );
```

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/13c92a89a69fc16f3d57af80fd527888c6f42248

oandregal commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/domain-field.tsx:
```diff
@@ -0,0 +1,83 @@
+			{ showPrimaryDomainLabel && <PrimaryDomainLabel /> }
+			<div className="domains-dataviews__domain-name">{ partialDomain.domain }</div>
+
+			{ domainTypeText && (
```
When you can use the latest DataViews (see state of [this PR](https://github.com/Automattic/wp-calypso/pull/96470)) this is a candidate to 1) become its own field and 2) become the [descriptionField](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#properties) to be rendered together with the `titleField` (what is now `primaryField`).

oandregal commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/domain-field.tsx:
```diff
@@ -0,0 +1,83 @@
+
+	return (
+		<Button className="domains-dataviews__domain-name-button" onClick={ onDomainClick }>
+			{ showPrimaryDomainLabel && <PrimaryDomainLabel /> }
```
When this label is rendered the field becomes too crowded. Shared a screenshot privately: p1737710599858909/1737614237.091909-slack-C082KU0C0MR

merkushin replied:
I think you forced showing this label as it is not supposed to be shown in this context :)
But, it just copies the appearance of the cell from DomainsTable. If you force showing it there, here is what you see:
![CleanShot 2025-01-25 at 00 18 03@2x](https://github.com/user-attachments/assets/d950187a-fd31-42a0-88d1-c18fb13f2fe6)

We had a brief discussion about if we needed to bring stuff like that as it has only one usage currently, and decided to do if there is no cost for it.

oandregal commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/domain-field.tsx:
```diff
@@ -0,0 +1,83 @@
+	};
+
+	return (
+		<Button className="domains-dataviews__domain-name-button" onClick={ onDomainClick }>
```
When updating to latest dataviews, this button can be removed in favor of the `onClickItem` [DataViews property](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#properties).

oandregal commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/domain-field.tsx:
```diff
@@ -0,0 +1,83 @@
+	domain: partialDomain,
+	isAllSitesView,
+	selectedFeature,
+	openDomainPane,
```
What's this property? I don't see it defined anywhere ([here](https://github.com/Automattic/wp-calypso/blob/542102d8cda107fd8642044650b271a53b55cca7/client/my-sites/domains/domain-management/list/bulk-all-domains.tsx#L682)).

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/b169ca7454da0519f160c63000652e01ceb94534

oandregal commented on client/my-sites/domains/domain-management/dataviews/dataviews-fields/ssl-status-field.tsx:
```diff
@@ -0,0 +1,37 @@
+			domainManagementLink={ domainManagementLink }
+			sslStatus={ domain.sslStatus }
+			hasWpcomManagedSslCert={ hasWpcomManagedSslCert }
+			as="div"
```
Thanks, it's important we generate valid HTML markup.

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+					return false;
+				}
+				return (
+					domain.currentUserIsOwner &&
```
This rule is not present in the production action ([see](https://github.com/Automattic/wp-calypso/blob/trunk/packages/domains-table/src/domains-table/domains-table-row-actions.tsx#L68)). As a result, domains that have that action in production, don't have it in the dataviews-powered version.

merkushin replied:
Yes, it is a bit complicated here.
It appeared here because it's also part of a bulk actions condition for the same action.
Initially, I didn't add it and I encountered cases when the action was available, but on the end screen I got a message that I'm not eligible for editing the contact information.
I just checked: the same happens with Domains Table.
![CleanShot 2025-01-24 at 23 07 10@2x](https://github.com/user-attachments/assets/cf77b35b-c955-4690-8441-ff23273f93ab)
So, I think having this condition here is a safer position. What do you think?

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+
+	const actions: Action< PartialDomainData >[] = [
+		{
+			id: 'manage-domain',
```
This action shouldn't be available in the list layout. We should aim to use the same patterns across wordpress.com, note [how it's implemented in wordpress.com/sites](https://github.com/Automattic/wp-calypso/blob/trunk/client/sites/components/sites-dataviews/actions.tsx#L328).

| Table | List |
| --- | --- |
| <img width="1685" alt="Screenshot 2025-01-24 at 11 14 25" src="https://github.com/user-attachments/assets/ecd57a4b-87cc-440e-aa7d-a3f85a8d5eea" /> | <img width="384" alt="Screenshot 2025-01-24 at 11 14 45" src="https://github.com/user-attachments/assets/34a45cba-674b-45be-9d27-305fac144da8" /> |

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/1244ab18e214695fdc68b1a5d58277a67b676f25

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+			},
+		},
+		{
+			id: 'manage-dns-settings',
```
For the same domain, in production, this action opens the list view with DNS section, e.g.: `https://wordpress.com/domains/manage/all/overview/<DOMAIN>/dns/<OTHER_DOMAIN>` while this PR opens calypsofied wp-admin (e.g.: `http://calypso.localhost:3000/home/<DOMAIN>`).

Why the difference?

merkushin replied:
It is a consequence of multiple conflict merges and unaligned changes (this PR lives too long).
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/8340320081bd5e4264d5ed81d52459489add39b2

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+		handleUpdateContactInfo,
+	} = useDomainsDataViewsContext();
+
+	const actions: Action< PartialDomainData >[] = [
```
This is general feedback that applies to all actions: 

A lot of the actions just open the list view with preview (settings), focused on a specific section (DNS), e.g.: `https://wordpress.com/domains/manage/all/overview/<DOMAIN>/dns/<OTHER_DOMAIN>`. This is confusing and dotcom design is trying to improve this experience for other parts of the product. In the meantime, can we simplify all of this by having a single action that opens the settings panel? This is: we never open a specific section such as DNS. I think this would be an improvement while we have for the larger redesign.

merkushin replied:
@Imran92 do you have a strong opinion about it in the context of the Domain Management revamp project?

Imran92 replied:
I don't have any strong opinion about this. The Domain Management Revamp project is wrapped up, so we can create an issue to be picked up later separately from the project.

I just want to add @fditrapani in the loop to get his opinion on removing the other actions and keeping just one.

fditrapani replied:
I agree with @Imran92 that we should leave it as is for now and create a follow up issue to review alongside the other dashboards. 

From what I can tell, there are actions that take you deep into the site on the `/sites`. Like settings as an example. The current behaviour mirrors that pattern.

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+				}
+			},
+			label: translate( 'Manage contact information' ),
+			supportsBulk: ! sidebarMode,
```
I understand this is trying to enable/disable bulk support depending on what layout is displayed (table or list)? We don't need that, the layouts themselves handle that gracefully (list layout doesn't support bulk actions yet).

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/1244ab18e214695fdc68b1a5d58277a67b676f25

oandregal commented on client/my-sites/domains/domain-management/dataviews/use-actions.tsx:
```diff
@@ -0,0 +1,259 @@
+			},
+			callback: ( domains: Array< PartialDomainData > ) => {
+				const domain = domains.length > 0 && domains[ 0 ] && getFullDomain( domains[ 0 ] );
+				if ( ! domain ) {
```
We're re-rendering the whole `DotcomDomainsDataViews` component, which makes actions work (eligibility) even if we have partial data: in a later re-render they'll have the full data. The fact that this works doesn't mean is a good approach. I see two issues:

- Performance.
- User actions are not respected. Try clicking a domain before the full data is loaded: the settings panel won't open.

I'm sharing some screenshots/videos privately p1737716400982189/1737614237.091909-slack-C082KU0C0MR

oandregal commented on client/my-sites/domains/domain-management/dataviews/style.scss:
```diff
@@ -0,0 +1,198 @@
+			overflow: hidden;
+			padding-bottom: 0;
+
+			.dataviews-view-list__media-wrapper {
```
What's this for? I removed it and don't notice any change. We shouldn't add any CSS that targets internal DataViews classes, that's fragile.

merkushin replied:
Hum, I need to check. Maybe the DataViews library was updated?
The purpose of it was to hide a thumbnail, which was impossible to do otherwise in the prior version.

merkushin replied:
Still makes sense:
![CleanShot 2025-01-25 at 00 11 13@2x](https://github.com/user-attachments/assets/7a606906-3d72-4278-b9c1-376474207bea)

I know that in a newer version there is a view option for that. But with current verion it doesn't work.

oandregal commented on client/my-sites/domains/domain-management/dataviews/style.scss:
```diff
@@ -0,0 +1,198 @@
+				}
+			}
+
+			.dataviews-pagination {
```
Same as above: we shouldn't target internal DataViews classes.

merkushin replied:
Fixed here: https://github.com/Automattic/wp-calypso/pull/97677/commits/74dd186b841302a0b4493c0077cb055db48059e1

merkushin committed:
Fix minor issues: useMemo dependency and domain.owner check

merkushin committed:
Rename useView to initializeViewState and move useState to DomainsDataViews

merkushin committed:
Eliminate unused openDomainPane

merkushin committed:
Don't show the primary action in the list layout

merkushin committed:
Fix Manage DNS action

merkushin committed:
Don't apply styling for DataViews internals

merkushin commented:
Thank you for a thorough review @oandregal!
I addressed most comments, a couple left for a further discussion.
And I'll open a separate PR for the approach for preloading all data first, and showing the DataViews only after that.

oandregal committed:
Update primaryField to titleField

oandregal committed:
Use onClickItem instead of implementing click behaviour at the field level

oandregal committed:
Domain connection as descriptionField

oandregal committed:
Update selection

oandregal commented:
@merkushin the WordPress package update landed yesterday, which includes the latest DataViews version. I thought I'd help move this PR forward and so I pushed the required updates to make it work with the last version.

merkushin commented:
@oandregal Thanks for updating it! I noticed one issue which I suppose might be related to `onClickItem`: if I click on a checkbox my click works as I click on the row, so instead of ticking the item it opens the pane.

merkushin committed:
Update styles from trunk

merkushin committed:
Remove unnecessary styling

merkushin committed:
Allow using checkboxes for bulk actions

oandregal committed:
Manage change selection in list view

oandregal commented:
@merkushin I've pushed a https://github.com/Automattic/wp-calypso/pull/97677/commits/49169454308c4de6222971f11746e55657223b02 that reverts https://github.com/Automattic/wp-calypso/pull/97677/commits/fdf536f82a313954d991725888ec385fe5757ebf and updates this condition: `if ( items.length === 1 ) {` to `if ( view.type === 'list' ) {`. This was an oversight on my part when updating to latest dataviews version, so sorry for the confusion.

Video with details at p1738917034753299-slack-C082KU0C0MR

merkushin committed:
Fix styles after rebasing

merkushin committed:
Fix pagination: use actual page for onClickItem

ouikhuan reviewed (commented):
Great work overall @merkushin ! This time I focused on functionality tests, I noticed there were a few border missing compare to others like `/sites`. For example, the borders on the sidebar and on the main panel is disappeared compared to others.

<img width="1445" alt="Screenshot 2025-02-11 at 3 53 54 PM" src="https://github.com/user-attachments/assets/ed265c78-5990-4758-942c-6f0f1369bdf3" />

<img width="443" alt="Screenshot 2025-02-11 at 3 54 03 PM" src="https://github.com/user-attachments/assets/6f88d38c-1f6c-412c-ade9-f6b1a05fa12a" />

Also, when you click on "Manage contact information"  on a single domain, current production version takes you to `https://wordpress.com/domains/manage/all/contact-info/edit/DOMAIN/DOMAIN`, but the behavior in the PR takes you to `http://calypso.localhost:3000/domains/manage/DOMAIN/edit-contact-info/DOMAIN` . Is this intentional?

phcp reviewed (commented):
Great work! Tested the dataview features, and it works nicely. The only issue I noticed was related to the Editing contact information (not related to the PR itself). When clicking to edit contact for one domain, it goes to the old UI, and when using bulk actions, it goes to a broken version of the page (mixing new and old layouts).

<img width="500" alt="Screenshot 2025-02-11 at 16 08 13" src="https://github.com/user-attachments/assets/0152fd55-2ecc-434f-8e77-d87249d8f0d3" />

merkushin committed:
Restore a border separating header from content

merkushin committed:
Fix contact info action url for a single domain

merkushin committed:
Resolve styling conflicts

Imran92 reviewed (commented):
I just noticed that we loose the sorting, pagination and all other data after tab switching and also closing the flyout pane. I'm sending the attachment in DM

merkushin committed:
Preserve query params in overview pane tabs

merkushin commented:
Thanks @Imran92! I just pushed the fix for it: https://github.com/Automattic/wp-calypso/pull/97677/commits/1167ce2856ce7ce9c93a7980ca0f29b91565b524

merkushin committed:
Adjust tests for preserving query params in overview pane tabs

merkushin committed:
Preserve query params for closing the overview pane

merkushin committed:
Adjust tests for preserving query params when closing overview pane

ouikhuan reviewed (approved):
The issue mentioned above has all been addressed, LGTM 🚢 

The only thing to me is that the query params also preserved on the original domain list now.

Imran92 reviewed (commented):
LGTM too 👍 

Another small issue I've noticed is the action menu icons are always visible. They should be visible only on hover. Similar to /sites -

<img width="207" alt="image" src="https://github.com/user-attachments/assets/055cd32f-ea41-4627-9948-599166ab62b4" />

merkushin committed:
Fix container height and padding

merkushin committed:
Merge branch 'trunk' into add/domains-data-views
